std/math.d

```diff
@@ -6798,3 +6799,205 @@ real yl2xp1(real x, real y) @nogc @safe pure nothrow;       // y * log2(x + 1)
+            return 1;
+        else
+            return 0;
+    }
+    else
+    {
+        // IBM Extended doubledouble does not follow the general
+        // sign-exponent-significand layout, so has to be handled generically
+
+        int xSign = signbit(x),
+            ySign = signbit(y);
+
+        if (xSign == 1 && ySign == 1)
+            return cmp(-y, -x);
+        else if (xSign == 1)
+            return -1;
+        else if (ySign == 1)
+            return 1;
+        else if (x < y)
+            return -1;
+        else if (x == y)
+            return 0;
+        else if (x > y)
+            return 1;
+        else if (isNaN(x) && !isNaN(y))
+            return 1;
+        else if (isNaN(y) && !isNaN(x))
+            return -1;
+        else if (getNaNPayload(x) < getNaNPayload(y))
+            return -1;
+        else if (getNaNPayload(x) > getNaNPayload(y))
+            return 1;
+        else
+            return 0;
+    }
+}
+
+/// Most numbers are ordered naturally.
+unittest
+{
+    assert(cmp(-double.infinity, -double.max) == -1);
+    assert(cmp(-double.max, -100.0) == -1);
+    assert(cmp(-100.0, -0.5) == -1);
+    assert(cmp(-0.5, 0.0) == -1);
+    assert(cmp(0.0, 0.5) == -1);
+    assert(cmp(0.5, 100.0) == -1);
+    assert(cmp(100.0, double.max) == -1);
+    assert(cmp(double.max, double.infinity) == -1);
+
+    assert(cmp(1.0, 1.0) == 0);
+}
+
+/// Positive and negative zeroes are distinct.
+unittest
+{
+    assert(cmp(-0.0, +0.0) == -1);
+    assert(cmp(+0.0, -0.0) == 1);
+}
+
+/// Depending on the sign, $(NAN)s go to either end of the spectrum.
+unittest
+{
+    assert(cmp(-double.nan, -double.infinity) == -1);
+    assert(cmp(double.infinity, double.nan) == -1);
+    assert(cmp(-double.nan, double.nan) == -1);
+}
+
+/// $(NAN)s of the same sign are ordered by the payload.
+unittest
+{
+    assert(cmp(NaN(10), NaN(20)) == -1);
+
+    assert(cmp(-NaN(20), -NaN(10)) == -1);
+}
+
+unittest
+{
+    import std.typetuple;
+    foreach (T; TypeTuple!(float, double, real))
+    {
+        T[] values = [-cast(T)NaN(20), -T.nan, -T.infinity, -T.max, -T.max / 2,
+                      T(-16.0), T(-1.0).nextDown, T(-1.0), T(-1.0).nextUp,
+                      T(-0.5), -T.min_normal, (-T.min_normal).nextUp,
+                      -2 * T.min_normal * T.epsilon,
+                      -T.min_normal * T.epsilon,
+                      T(-0.0), T(0.0),
+                      T.min_normal * T.epsilon,
+                      2 * T.min_normal * T.epsilon,
+                      T.min_normal.nextDown, T.min_normal, T(0.5),
+                      T(1.0).nextDown, T(1.0), T(1.0).nextUp, T(16.0),
+                      T.max / 2, T.max, T.infinity, T.nan, cast(T)NaN(20)];
+
+        foreach (i, x; values[0 .. $ - 1])
+        {
+            foreach (y; values[i + 1 .. $])
+            {
+                assert(cmp(x, y) == -1);
+                assert(cmp(y, x) == 1);
+            }
+            assert(cmp(x, x) == 0);
+        }
+    }
+}
```

Comment:
This function should work when `x` and `y` have different floating point types. Please add a unittest.